staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go

```diff
@@ -0,0 +1,294 @@
+		versionExists := false
+
+		// Not very efficient, but in practice there are generally not many versions
+		for i := range existing.Versions {
+			if existing.Versions[i].Version == value.Version {
+				existing.Versions[i] = value
+				versionExists = true
+				break
+			}
+		}
+
+		if !versionExists {
+			existing.Versions = append(existing.Versions, value)
+		}
+
+	} else {
+		rdm.apiGroups[groupName] = &metav1.APIGroupDiscovery{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "APIGroupDiscovery",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: groupName,
+			},
+			Versions: []metav1.APIVersionDiscovery{value},
+		}
+		rdm.apiGroupNames = append(rdm.apiGroupNames, groupName)
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroupVersion(apiGroup metav1.GroupVersion) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	group, exists := rdm.apiGroups[apiGroup.Group]
+	if !exists {
+		return
+	}
+
+	for i := range group.Versions {
+		if group.Versions[i].Version == apiGroup.Version {
+			group.Versions = append(group.Versions[:i], group.Versions[i+1:]...)
+			break
+		}
+	}
+
+	if len(group.Versions) == 0 {
+		delete(rdm.apiGroups, group.Name)
+		for i := range rdm.apiGroupNames {
+			if rdm.apiGroupNames[i] == group.Name {
+				rdm.apiGroupNames = append(rdm.apiGroupNames[:i], rdm.apiGroupNames[i+1:]...)
+				break
+			}
+		}
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroup(groupName string) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	delete(rdm.apiGroups, groupName)
+	for i := range rdm.apiGroupNames {
+		if rdm.apiGroupNames[i] == groupName {
+			rdm.apiGroupNames = append(rdm.apiGroupNames[:i], rdm.apiGroupNames[i+1:]...)
+			break
+		}
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) WebService() *restful.WebService {
+	mediaTypes, _ := negotiation.MediaTypesForSerializer(rdm.serializer)
+	ws := new(restful.WebService)
+	ws.Path(DiscoveryEndpointRoot)
+	ws.Doc("get available API groupversions and resources")
+
+	ws.Route(ws.GET("/v2").To(func(req *restful.Request, resp *restful.Response) {
+		rdm.ServeHTTP(resp.ResponseWriter, req.Request)
+	}).
+		Doc("get available API groupversions and their resources").
+		Operation("getDiscoveryResources").
+		Produces(mediaTypes...).
+		Consumes(mediaTypes...).
+		Writes(metav1.APIGroupDiscoveryList{}))
+	return ws
+}
+
+func (rdm *resourceDiscoveryManager) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	rdm.lock.RLock()
+	response := rdm.cachedResponse
+	etag := rdm.cachedResponseETag
+
+	// The cachedResponse is wiped out every time there might be a change to its
+	// contents.
+	if response != nil {
+		defer rdm.lock.RUnlock()
```

Comment:
The need to hold a readlock while `ServeHTTPWithETag` isn't obvious as I read top to bottom here.  Are we protecting something in `ServeHTTPWithETag` or are we just protecting the response/etag reads above this block?